lib/db/observed.go

```diff
@@ -0,0 +1,253 @@
+			deviceID := protocol.DeviceIDFromBytes(keyDev)
+			if device != protocol.EmptyDeviceID && device != deviceID {
+				continue
+			}
+			var of ObservedFolder
+			folderID := string(db.keyer.FolderFromPendingFolderKey(iter.Key()))
+			if len(folderID) < 1 {
+				goto deleteKey
+			}
+			if err := of.Unmarshal(bs); err != nil {
+				goto deleteKey
+			}
+			if _, ok := res[folderID]; !ok {
+				res[folderID] = make(map[protocol.DeviceID]ObservedFolder)
+			}
+			res[folderID][deviceID] = of
+			continue
+		}
+	deleteKey:
+		l.Warnln("Invalid pending folder entry, deleting from database")
+		if err := db.Delete(iter.Key()); err != nil {
+			return nil, err
+		}
+	}
+	return res, nil
+}
+
+// RemovePendingFolder removes entries for specific folder / device combinations
+func (db *Lowlevel) RemovePendingFolder(id string, devices []protocol.DeviceID) {
+	for _, dev := range devices {
+		key, err := db.keyer.GeneratePendingFolderKey(nil, []byte(id), dev[:])
+		if err != nil {
+			continue
+		}
+		if err := db.Delete(key); err != nil {
+			l.Warnf("Failed to remove pending folder entry: %v", err)
+		}
+	}
+}
+
+// Set of devices for which pending folders are allowed, but not specific folder IDs
+type DropListObserved map[protocol.DeviceID]map[string]bool
+
+func (dl DropListObserved) MarkDevice(device protocol.DeviceID) map[string]bool {
+	folders, ok := dl[device]
+	if !ok {
+		dl[device] = nil
+	}
+	return folders
+}
+
+func (dl DropListObserved) MarkFolder(folder string, devices []protocol.DeviceID, myID protocol.DeviceID) {
+	for _, dev := range devices {
+		if dev == myID {
+			continue
+		}
+		folders := dl.MarkDevice(dev)
+		if folders == nil {
+			folders = make(map[string]bool)
+			dl[dev] = folders
+		}
+		folders[folder] = true
+	}
+}
+
+// CleanPendingDevices removes all pending device entries matching a given set of device IDs
+func (db *Lowlevel) CleanPendingDevices(dropList DropListObserved) error {
+	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingDevice})
+	if err != nil {
+		l.Warnf("Could not iterate through pending device entries for cleanup: %v", err)
+		return err
+	}
+	defer iter.Release()
+	for iter.Next() {
+		_, err := db.Get(iter.Key())
+		if err != nil {
+			return err
+		}
+		keyDev := db.keyer.DeviceFromPendingDeviceKey(iter.Key())
+		//FIXME: DeviceIDFromBytes() panics when given a wrong length input.
+		//       It should rather return an error which we'd check for here.
+		if len(keyDev) == protocol.DeviceIDLength {
+			// Valid entries are looked up in the drop-list, invalid ones cleaned up
+			deviceID := protocol.DeviceIDFromBytes(keyDev)
+			_, dropDev := dropList[deviceID]
+			if !dropDev {
+				continue
+			}
+			l.Debugf("Removing marked pending device %v", deviceID)
+		} else {
+			l.Warnln("Invalid pending device entry, deleting from database: %v", keyDev)
+		}
+		if err := db.Delete(iter.Key()); err != nil {
+			l.Warnf("Failed to remove pending device entry: %v", err)
```

Comment:
In this very situation it would be complicated to pass the error to the caller, since it's only one loop iteration, while the others might succeed.  Returning multiple errors in a slice is overkill, so I log it here and try to continue with the other records.

I'll improve on the error messages and remove the return value probably.  There is only two users of this API (both in model) and both would get uglier if moving the logging over there.

In general: This whole thing is about discarding unwanted stuff.  If that fails (very unlikely), there must be some deeper reason and all we can do is give the maintainters good log hints for debugging.  Context is also not important for these errors I'd say.